### PR TITLE
Add Deferred and PromiseMap dependencies

### PR DESCRIPTION
**Pull request by yungsters**
Pulls in `Deferred` and `PromiseMap`. This also required adding `promise` as a package dependency and adding it to the `moduleMap` config.

**Test Plan:**

```
npm test
gulp
```

### CODE REVIEW

**Comment by yungsters** [comment]
We should try to consolidate the Babel configuration that currently exists in `preprocessor.js` and `gulpfile.js`.

**Comment by yungsters** [comment]
Actually, Flow won't be happy if we don't have a real `Promise` module, so let me fix this.

File: src/core/Deferred.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright 2004-present Facebook. All Rights Reserved.
```
Comment (zpao) [inline src/core/Deferred.js]: We should make sure this header is updated

**Comment by zpao** [comment]
> We should try to consolidate the Babel configuration that currently exists in preprocessor.js and gulpfile.js.

[My comment](https://github.com/facebook/fbjs/blob/master/scripts/jest/preprocessor.js#L1-L2) agrees with you :)

**Comment by zpao** [comment]
accepting because I know you'll fix the license headers

**Comment by yungsters** [comment]
> accepting because I know you'll fix the license headers

So much faith. ;]

**Pull request merged**